app/src/main/java/net/gsantner/markor/format/zimwiki/ZimWikiReplacePatternGenerator.java

```diff
@@ -0,0 +1,121 @@
+
+        return patterns;
+    }
+
+    private TextActions.ReplacePattern removeHeadingCharsForExactHeadingLevel(String headingChars) {
+        return new TextActions.ReplacePattern(
+                "^\\s{0,3}" + headingChars + "[ \\t](.*?)[ \\t]" + headingChars + "\\w*",
+                "$1");
+    }
+
+    private TextActions.ReplacePattern replaceDifferentHeadingLevelWithThisLevel(String headingChars) {
+        return new TextActions.ReplacePattern("^\\s{0,3}={0,6}([ \\t].*?[ \\t])={0,6}",
+                headingChars+"$1"+headingChars);
+    }
+
+    private TextActions.ReplacePattern createHeadingIfNoneThere(String headingChars) {
+        return new TextActions.ReplacePattern("^\\s*?(\\S?.*)\\s*",
+                headingChars+" $1 "+headingChars);
+    }
+
+    public List<TextActions.ReplacePattern> replaceWithNextStateCheckbox() {
+        List<TextActions.ReplacePattern> replacePatterns = new ArrayList<>();
+
+        // toggle order: no checkbox -> unchecked -> checked -> crossed -> arrow -> unchecked -> ...
+        replacePatterns.addAll(toggleCheckboxToNextState());
+        replacePatterns.addAll(replaceOtherPrefixesWithUncheckedBox());
+        return replacePatterns;
+    }
+
+    private List<TextActions.ReplacePattern> toggleCheckboxToNextState() {
+        return Arrays.asList(
+                new TextActions.ReplacePattern(PREFIX_UNCHECKED_LIST, checkedReplacement),
+                new TextActions.ReplacePattern(PREFIX_CHECKED_LIST, crossedReplacement),
+                new TextActions.ReplacePattern(PREFIX_CROSSED_LIST, arrowReplacement),
+                new TextActions.ReplacePattern(PREFIX_ARROW_LIST, uncheckedReplacement));
+    }
+
+    private List<TextActions.ReplacePattern> replaceOtherPrefixesWithUncheckedBox() {
+        List<TextActions.ReplacePattern> replacePatterns = new ArrayList<>();
+        for (final Pattern otherPattern : PREFIX_PATTERNS) {
+            replacePatterns.add(new TextActions.ReplacePattern(otherPattern, uncheckedReplacement));
+        }
+        return replacePatterns;
+    }
+
+    public List<TextActions.ReplacePattern> replaceWithUnorderedListPrefixOrRemovePrefix() {
+        return replaceOtherPrefixWithSelectedOrRemovePrefix(PREFIX_PATTERNS, PREFIX_UNORDERED_LIST, unorderedListReplacement);
+    }
+
+    public List<TextActions.ReplacePattern> replaceWithOrderedListPrefixOrRemovePrefix() {
+        return replaceOtherPrefixWithSelectedOrRemovePrefix(PREFIX_PATTERNS, PREFIX_ORDERED_LIST, orderedListReplacement);
+    }
+
+    public List<TextActions.ReplacePattern> deindentOneTab() {
+        final Pattern leadingTabIndentPattern = Pattern.compile("^\t");
+        return Collections.singletonList(new TextActions.ReplacePattern(leadingTabIndentPattern, ""));
+    }
+
+    public List<TextActions.ReplacePattern> indentOneTab() {
+        final Pattern lineStart = Pattern.compile("^");
+        return Collections.singletonList(new TextActions.ReplacePattern(lineStart, "\t"));
```

Comment:
If making lineStart etc static final at the class level is not possible, then `TextActions.ReplacePattern` has a constructor which will take 2 strings and call pattern.compile internally. No need to do this in 2 steps. Also, `regexReplaceAction` has an overload [which should do all this for you](https://github.com/gsantner/markor/blob/master/app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java#L356-L358).

Reply:
Changed this to use the two-string constructor.

The other thing is probably a matter of taste, but in some cases (like here) I prefer longer but more obvious code over shorter one. Especially when it comes to meaningful method and variable names. So instead of just using the overload and calling `regexReplaceAction("^\t", "")` directly in `ZimWikiTextActions`, the method `deindentOneTab` states much more clearly what the intended effect of the action is.